olive/common/ort_inference.py

```diff
@@ -51,26 +71,313 @@ def get_ort_inference_session(
+        return [], []
+
+    provider_name_to_options = collections.OrderedDict()
+
+    def set_provider_options(name, options):
+        if available_provider_names and name not in available_provider_names:
+            logger.warning(
+                "Specified provider '%s' is not in available provider names.Available providers: '%s'",
+                name,
+                ", ".join(available_provider_names),
+            )
+
+        if name in provider_name_to_options:
+            logger.warning("Duplicate provider '%s' encountered, ignoring.", name)
+            return
+
+        normalized_options = {str(key): str(value) for key, value in options.items()}
+        provider_name_to_options[name] = normalized_options
+
+    if not isinstance(providers, collections.abc.Sequence):
+        raise ValueError("'providers' should be a sequence.")
+
+    if provider_options is not None:
+        if not isinstance(provider_options, collections.abc.Sequence):
+            raise ValueError("'provider_options' should be a sequence.")
+
+        if len(providers) != len(provider_options):
+            raise ValueError("'providers' and 'provider_options' should be the same length if both are given.")
+
+        if not all(isinstance(provider, str) for provider in providers):
+            raise ValueError("Only string values for 'providers' are supported if 'provider_options' is given.")
+
+        if not all(isinstance(options_for_provider, dict) for options_for_provider in provider_options):
+            raise ValueError("'provider_options' values must be dicts.")
+
+        for name, options in zip(providers, provider_options):
+            set_provider_options(name, options)
+    else:
+        for provider in providers:
+            if isinstance(provider, str):
+                set_provider_options(provider, {})
+            elif (
+                isinstance(provider, (tuple, list))
+                and len(provider) == 2
+                and isinstance(provider[0], str)
+                and isinstance(provider[1], dict)
+            ):
+                set_provider_options(provider[0], provider[1])
+            else:
+                raise ValueError("'providers' values must be either strings or (string, dict) tuples.")
+
+    return list(provider_name_to_options.keys()), list(provider_name_to_options.values())
+
+
+def check_ort_fallback(session: "InferenceSession", execution_providers: Sequence[str]):
+    """Check if the onnxruntime fallback happens and raise an error if it does."""
+    # pylint: disable=protected-access
+    session_providers = session.get_providers()
+    for ep in execution_providers:
+        if ep not in session_providers:
+            raise RuntimeError(
+                f"The onnxruntime fallback happens. {ep} is not in the session providers {session_providers}."
+                f" session._enable_fallback = {session._enable_fallback}"
+            )
+    session.disable_fallback()
+
+
+def run_inference(
```

Comment:
Will replace `run_inference` and `time_inference` with a InferenceSession wrapper with io_binding. Similar to https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/io_binding_helper.py#L211. 

The common code we want to abstract is the io_binding related code. `run_inference` is hacky with the requirements for data_generator and post_run parameters. We can do the looping in the evaluator.